std-bits/base/src/main/java/org/enso/base/parser/NumberWithSeparators.java

```diff
@@ -0,0 +1,348 @@
+  UNKNOWN(Constants.UNKNOWN, Constants.UNKNOWN),
+
+  // Special case where we have encountered a . with 3 trailing digits. ##0.123 ####.123
+  DOT_UNKNOWN(Constants.UNKNOWN, '.'),
+  // Special case where we have encountered a single . within 3 digits from start and without 3
+  // digits from end.
+  UNKNOWN_DOT(Constants.UNKNOWN, '.'),
+  // Special case where we have encountered a , with 3 trailing digits.
+  COMMA_UNKNOWN(',', Constants.UNKNOWN),
+  // Special case where we have encountered a single . within 3 digits from start and without 3
+  // digits from end.
+  UNKNOWN_COMMA(Constants.UNKNOWN, '.'),
+
+  NO_UNKNOWN(Constants.NONE, Constants.UNKNOWN),
+  NO_DOT(Constants.NONE, '.'),
+  NO_COMMA(Constants.NONE, ','),
+
+  DOT_COMMA('.', ','),
+
+  COMMA_DOT(',', '.'),
+
+  SPACE_UNKNOWN(' ', Constants.UNKNOWN),
+  SPACE_DOT(' ', '.'),
+  SPACE_COMMA(' ', ','),
+
+  SWISS_UNKNOWN('\'', Constants.UNKNOWN),
+  SWISS_DOT('\'', '.'),
+  SWISS_COMMA('\'', ',');
+
+  public static NumberWithSeparators fromSeparators(String thousand, String decimal) {
+    if (thousand != null && thousand.length() > 1) {
+      throw new IllegalArgumentException("Invalid thousand separator (more than one character).");
+    }
+
+    if (decimal != null && decimal.length() > 1) {
+      throw new IllegalArgumentException("Invalid decimal separator (more than one character).");
+    }
+
+    char thousands =
+        thousand == null
+            ? Constants.UNKNOWN
+            : (thousand.isEmpty() ? Constants.NONE : thousand.charAt(0));
+    char decimals =
+        decimal == null
+            ? Constants.UNKNOWN
+            : (decimal.isEmpty() ? Constants.NONE : decimal.charAt(0));
+
+    // For these special cases if we know the decimal separator should fix it.
+    if (thousands == Constants.UNKNOWN) {
+      if (decimals == '.') {
+        return UNKNOWN_DOT;
+      } else if (decimals == ',') {
+        return UNKNOWN_COMMA;
+      }
+    }
+
+    var matched =
+        Arrays.stream(values())
+            .filter(v -> v.thousands == thousands && v.decimal == decimals)
+            .findAny();
+    if (matched.isEmpty()) {
+      throw new IllegalArgumentException("Invalid separators.");
+    }
+    return matched.get();
+  }
+
+  static class Constants {
+    static final char NONE = '\0';
+    static final char UNKNOWN = '\uFFFD';
+  }
+
+  static boolean isDigit(char c) {
+    return (c >= '0' && c <= '9');
+  }
+
+  private final char thousands;
+  private final char decimal;
+
+  NumberWithSeparators(char thousands, char decimal) {
+    this.thousands = thousands;
+    this.decimal = decimal;
+  }
+
+  public char getThousands() {
+    return thousands;
+  }
+
+  public char getDecimal() {
+    return decimal;
+  }
+
+  NumberParseResult parse(CharSequence value, int idx, boolean integer) {
+    var separators = Separators.parse(value, idx, integer);
+    if (separators == null) {
+      return new NumberParseFailure("Invalid separators.");
+    }
```

Comment:
For later: to get more friendly error reporting it would be great to save somewhere (perhaps in a `SeparatorParseResult`) what character was encountered to cause the 'invalid separators'.

Reply:
Added to #11228.